### PR TITLE
Cache Digest constants

### LINKED ISSUE #13578: SHA performance test
When I compared openssl 3.0 and openssl 1.1.1f, I found that the performance of alpha9 was lower than that of alpha5.
I am also comparing with OpenSSL 1.1.1f. I wonder if this can explain the problem.

All numbers are Percentage Diff (%)(alpha5-1.1.1f/1.1.1f)
alpha5 | 16 bytes | 64 bytes | 256 bytes | 1024 bytes | 8192 bytes | 16384 bytes
-- | -- | -- | -- | -- | -- | --
sha1 | -74.24694646 | -67.83789878 | -56.77331253 | -20.98058948 | -4.089929654 |-2.746844537
sha256 |  -68.18784269 | -59.0421308 | -42.57797189 | -20.2773497 | 2.462470535 | -14.52529967
sha512 |  -56.98251239 | -56.75811124 | -44.69438411 | -19.60932482 | 2.45562681 | -33.7214586

All numbers are Percentage Diff (%)(alpha9-1.1.1f/1.1.1f)
alpha9 | 16 bytes | 64 bytes | 256 bytes | 1024 bytes | 8192 bytes | 16384 bytes
-- | -- | -- | -- | -- | -- | --
sha1 |  -74.93675632 | -68.51181924 |  -55.35508324 |  -35.25122672 | -5.811513632 |  -2.937482437
sha256 | -93.95017321 |  -93.97662812 |  -88.74451967 |  -77.98002727 | -51.3641645 |  -46.51934182
sha512 |  -89.04846118 |  -93.23677374 | -84.93249856 | -76.25703978 |  -24.29898999 |  -37.92422949

using `openssl speed -evp` ,Five tests, and the results are almost the same.

### PR DESCRIPTION
**Pull request by mattcaswell**
`EVP_CIPHER` already caches certain constants so that we don't have to query the provider every time. We do the same thing with `EVP_MD` constants. Without this we can get performance issues, e.g. running "speed" with small blocks of data to digest can spend a long time in `EVP_MD_size()`, which should be quick.

Partialy fixes #13578

I've also refactored the `EVP_CIPHER` constant caching to move it into `evp_cipher_from_dispatch`, i.e. to where the `EVP_CIPHER` actually gets constructed rather than where it is fetched in `EVP_CIPHER_fetch`. The problem with the latter is that we construct a cipher once, but we may fetch it many times. We shouldn't have to cache the constants every time we fetch.

### CODE REVIEW

**Comment by mattcaswell** [comment]
I've run some tests to see the effect on performance of this change. I ran both the "before" and "after" tests with #13721 applied. I'm seeing quite a lot of variability in the results from one run to the next, so my confidence the precise numbers is quite low. To try and reduce variability as much as possible I "warmed up" my machine before both the "before" and "after" tests by repeating the same speed command 5 times and discarding the results. I then ran the command again 5 times and recorded the results.

I compiled with just `config --strict-warnings`, and ran this speed command:

````
openssl speed -evp sha256 -bytes 16
````
The "before" output was:
````
sha256           38566.15k
sha256           36042.71k
sha256           42937.12k
sha256           42312.67k
sha256           44117.60k
````
Which gives an average of 40795.25k

The "after" output was:
````
sha256           46979.81k
sha256           49308.40k
sha256           48095.66k
sha256           46935.39k
sha256           46201.71k
````
Which gives an average of 47504.19k.

This represents a speed-up of approximately 16% (if you believe the numbers).

File: crypto/evp/evp_lib.c [hunk 1]
```diff
@@ -688,19 +688,7 @@ const OSSL_PROVIDER *EVP_MD_provider(const EVP_MD *md)
...
 int EVP_MD_block_size(const EVP_MD *md)
 {
-    int ok;
-    size_t v;
-    OSSL_PARAM params[2] = { OSSL_PARAM_END, OSSL_PARAM_END };
-
-    if (md == NULL) {
-        ERR_raise(ERR_LIB_EVP, EVP_R_MESSAGE_DIGEST_IS_NULL);
-        return -1;
-    }
```
Comment (beldmit) [inline crypto/evp/evp_lib.c:698]: Why do we don't check for md here but check in EVP_MD_size?

Comment (beldmit) [reply crypto/evp/evp_lib.c:698]: The same question is about EVP_MD_flags

Comment (mattcaswell) [reply crypto/evp/evp_lib.c:698]: I'm just reverting the implementation back to the way it used to be in 1.1.1. There, EVP_MD_size() has the check:

https://github.com/openssl/openssl/blob/64a1b940d2b640e5edf0feae90e81bbb6b4941e7/crypto/evp/evp_lib.c#L320-L327

But EVP_MD_block_size and EVP_MD_flags do not:
https://github.com/openssl/openssl/blob/64a1b940d2b640e5edf0feae90e81bbb6b4941e7/crypto/evp/evp_lib.c#L305-L308
https://github.com/openssl/openssl/blob/64a1b940d2b640e5edf0feae90e81bbb6b4941e7/crypto/evp/evp_lib.c#L329-L332

**Review by beldmit — APPROVED** [review]
LGTM

**Comment by kroeckx** [comment]
Deprecated functions like EVP_MD_meth_set_result_size() set md->md_size, which is now the cached value. Either they should actually change something in the provider, return an error, or be removed. This PR doesn't actually introduce the problem, but now seems to use existing variables that were not initialized before, but that we did read and then don't do anything with.

The functions are currently documented to just return 1.

**Comment by beldmit** [comment]
I think it's a separate issue and these functions should not be applicable to provided digests.

**Comment by mattcaswell** [comment]
> I think it's a separate issue and these functions should not be applicable to provided digests.

Exactly. Those functions are deprecated for a reason. They are intended for use when you are setting up your own custom EVP_MD which you have created yourself.

>  Either they should actually change something in the provider, return an error, or be removed. 

I disagree. They are deprecated and therefore their usage is strongly discouraged. Existing uses will not be broken. I don't see a problem here.

**Comment by beldmit** [comment]
Merged. Thanks!

**Pull request closed without merging**